### ISSUE TITLE
 Surface Laptop Studio Touchpad and Keyboard Randomly Freeze

### ISSUE THREAD

**Issue by m1nicrusher** [linux-surface/linux-surface#1026]
Fresh Install of Ubuntu Unity. Applied the quirk and it seems to be applied according to `libinput`.

`/etc/libinput/local-overrides.quirks`:
```
[Microsoft Surface Laptop Studio Touchpad]
MatchVendor=0x045E
MatchProduct=0x09AF
MatchUdevType=touchpad
AttrPressureRange=25:10
AttrPalmPressureThreshold=500
```

`sudo libinput quirks list /dev/input/event12` (event12 is listed as "Microsoft Surface 045E:09AF Touchpad")
```
AttrPressureRange=25:10
AttrPalmPressureThreshold=500
```

The bahaviour of the touchpad currently is:
- Can click
- Moving the cursor requires strong pressure
- Two finger scroll works (also requires strong pressure)

The quirk worked fine previously on Fedora.

### Environment

- Hardware model: Surface Laptop Studio
- Kernel version: 5.15.0-56-generic
- Distribution: Ubuntu Unity 22.04 LTS

**Comment by m1nicrusher** [comment]
Having a look of `libinput measure touchpad-pressure` reveals that the cursor will only move when the pressure is above around 230.

**Comment by qzed** [comment]
As far as I know, pretty much anything should use libinput, at least on Wayland. Is it possible that you're running XOrg/X11 and that somehow takes over handling of the touchpad? There should be a way to route that through the XOrg libinput driver (see https://www.mankier.com/4/libinput).

**Comment by onkoe** [comment]
Oh hey, I just made a Reddit post along these lines. https://new.reddit.com/r/SurfaceLinux/comments/1020g8w/surface_book_studio_sbs_trackpad_stops_working/

I'm having issues with longer drags on Fedora, no matter the screen resolution, scaling,  or display server.

**Comment by qzed** [comment]
@onkoe Can you check that the quirks for the touchpad are applied and also try out `libinput measure touchpad-pressure` to see if it's pressure related?

**Comment by onkoe** [comment]
@qzed I have that command a try. It seems that when it gets stuck, my "down" is often being detected as a "palm", too. 

I'm not fluent in libinput's tooling but this may be a part of the problem! Would love to see someone a bit more knowledgeable take a look on their SBS.

**Comment by qzed** [comment]
That would explain it. You can increase `AttrPalmPressureThreshold` in the custom entry in `/etc/libinput/local-overrides.quirks` (assuming you've followed the wiki entry and applied the quirk) until this doesn't happen any more.

**Comment by m1nicrusher** [comment]
> As far as I know, pretty much anything should use libinput, at least on Wayland. Is it possible that you're running XOrg/X11 and that somehow takes over handling of the touchpad? There should be a way to route that through the XOrg libinput driver (see https://www.mankier.com/4/libinput).

As far as I know Ubuntu Unity only supports Xorg. I don't even see an option to use Wayland. Anyway I'm switching back to Fedora as there are just too few people using this new flavour of Ubuntu. I guess we should deal with it later on when more people started to use it.

**Comment by m1nicrusher** [comment]
> That would explain it. You can increase `AttrPalmPressureThreshold` in the custom entry in `/etc/libinput/local-overrides.quirks` (assuming you've followed the wiki entry and applied the quirk) until this doesn't happen any more.

I had the same issue and I ended up setting the palm threshold to around 850. So far so good.

**Comment by onkoe** [comment]
Palm threshold of 850 also works really well for me. Does 500 work well for some people? If not, it may be worth updating [libinput's default quirk](https://gitlab.freedesktop.org/libinput/libinput/-/blob/main/quirks/30-vendor-microsoft.quirks) and the linux-surface wiki page.

**Comment by qzed** [comment]
IIRC the 500 comes from @Np-chaonay. If that works better for most then yes, we should update it.

**Comment by NP-chaonay** [comment]
> IIRC the 500 comes from @NP-chaonay. If that works better for most then yes, we should update it.

Sorry, but I not invent that value, it existed before I buy this SLS.

according to this, https://github.com/linux-surface/linux-surface/wiki/Surface-Laptop-Studio/b4b26a5fb0d437a43f4b64d450b314d5b6159ad1

someone have got the value before the time that above commit is done

**Comment by qzed** [comment]
Ah sorry, then I misremembered. I looked for the original thread: https://github.com/linux-surface/linux-surface/issues/596 and it seems it's actually been a suggestion from me based on some small data sample. So I guess we can change that without any expected problems.

Maybe anyone wants to give this some proper evaluation via `libinput measure touchpad-pressure` or is 850 already a good enough trade-off between palm detection going wrong and touches not being detected?

**Comment by m1nicrusher** [comment]
> Maybe anyone wants to give this some proper evaluation via `libinput measure touchpad-pressure` or is 850 already a good enough trade-off between palm detection going wrong and touches not being detected?

I didn't "invent" this value. I used the above command you mentioned and found that it showed around 850 when I was actually covering my palm on the touchpad. Plus, 850 really needs a very strong push using one finger.

Nevertheless it'll be very helpful to see other people's testing result.

**Comment by qzed** [comment]
Ah neat! Thanks for clarifying that. I agree that a couple of more results would be great to make sure this value works for everyone. @NP-chaonay?

**Comment by NP-chaonay** [comment]
> Ah neat! Thanks for clarifying that. I agree that a couple of more results would be great to make sure this value works for everyone. @NP-chaonay?

Sorry but how I should test and is anything I should know I about since I have not using related calibrate tool of that value.

Just check if 850 still have good experience?

@qzed

**Comment by qzed** [comment]
That, and you could use `libinput measure touchpad-pressure` to check the actual values (then you can probably test better with your palm etc.)

**Comment by m1nicrusher** [comment]
You can also use `sudo libinput measure touchpad-pressure --touch-thresholds=x:y --palm-threshold=n` to test before applying to your system config.

**Comment by m1nicrusher** [comment]
By the way, does anyone know why we need `touch-thresholds`? Shouldn't greater than one means you are touching the touchpad? Why say we need 25 to mark it as `down` and less than 10 to mark it as `up`?

**Comment by qzed** [comment]
From https://github.com/linux-surface/linux-surface/issues/596#issuecomment-939646383: The default thresholds seem to be 240 for touch-down, 200 for touch-up, and 130 for palm. So the default values to seem to be quite nonsensical. I'm not sure where libinput gets them from.

As for the actual values: Those are guesses by me again. I've just looked at the avg and median values and chose something that is well below them, but also a bit above zero. These values should again be a trade-off: They should be low enough to detect all deliberate touches and high enough to not cause any spurious touch inputs (I guess some electrostatic noise or other stuff could cause them if it's too low). So I think ideally you choose them as the highest value possible that retains all deliberate touch events.

I think it would be a good idea to test them as well and see if they need adjustments.

**Comment by m1nicrusher** [comment]
> From [#596 (comment)](https://github.com/linux-surface/linux-surface/issues/596#issuecomment-939646383): The default thresholds seem to be 240 for touch-down, 200 for touch-up, and 130 for palm. So the default values to seem to be quite nonsensical. I'm not sure where libinput gets them from.
> 

That probably explains the behaviour on Ubuntu Unity, which means the quirk doesn't get applied on the system. Maybe they use a different utility to control touchpad? I'm assuming their package structures are more similar to a mix-up of Linux Mint and Ubuntu 17/16.04 (when they still use Unity). Any idea how to allow Ubuntu Unity make use of the quirk?

> As for the actual values: Those are guesses by me again. I've just looked at the avg and median values and chose something that is well below them, but also a bit above zero. These values should again be a trade-off: They should be low enough to detect all deliberate touches and high enough to not cause any spurious touch inputs (I guess some electrostatic noise or other stuff could cause them if it's too low). So I think ideally you choose them as the highest value possible that retains all deliberate touch events.

Hmmm, that makes sense. Will do some test on that as well. I barely remember it stays absolute 0 whenever I'm not touching it but yeah, considering how touchpad works, electrostatic may affect it.
Thanks for the explanation!

**Comment by qzed** [comment]
> Any idea how to allow Ubuntu Unity make use of the quirk?

Maybe it's X11 related and explicitly needs libinput to be chosen as driver as I mentioned in https://github.com/linux-surface/linux-surface/issues/1026#issuecomment-1369803906? Apart from that, I have no idea.

**Comment by m1nicrusher** [comment]
> They should be low enough to detect all deliberate touches and high enough to not cause any spurious touch inputs (I guess some electrostatic noise or other stuff could cause them if it's too low). So I think ideally you choose them as the highest value possible that retains all deliberate touch events.

OK, I've done some test on my Surface Laptop Studio. Since I don't have Windows installed, I tested the behaviour of the touchpad in Surface UEFI. It seems to me that no matter how light your finger touches it, it will always detect as down. There seems to be no mechanism to prevent something like "spurious touch input". And, I tested it with some conductive objects (a key, the clip I used to short PWR pins on my desktop, a flash drive with metal USB port, a 3.5mm cable) and none of them seems to be able to trigger the touchpad. Maybe the touchpad doesn't work like a capacitive touchscreen? I've been using the quirk of 1:0:850 for a while and it seems just fine at the moment.

**Comment by NP-chaonay** [comment]
> > They should be low enough to detect all deliberate touches and high enough to not cause any spurious touch inputs (I guess some electrostatic noise or other stuff could cause them if it's too low). So I think ideally you choose them as the highest value possible that retains all deliberate touch events.
> 
> OK, I've done some test on my Surface Laptop Studio. Since I don't have Windows installed, I tested the behaviour of the touchpad in Surface UEFI. It seems to me that no matter how light your finger touches it, it will always detect as down. There seems to be no mechanism to prevent something like "spurious touch input". And, I tested it with some conductive objects (a key, the clip I used to short PWR pins on my desktop, a flash drive with metal USB port, a 3.5mm cable) and none of them seems to be able to trigger the touchpad. Maybe the touchpad doesn't work like a capacitive touchscreen? I've been using the quirk of 1:0:850 for a while and it seems just fine at the moment.



> > They should be low enough to detect all deliberate touches and high enough to not cause any spurious touch inputs (I guess some electrostatic noise or other stuff could cause them if it's too low). So I think ideally you choose them as the highest value possible that retains all deliberate touch events.
> 
> OK, I've done some test on my Surface Laptop Studio. Since I don't have Windows installed, I tested the behaviour of the touchpad in Surface UEFI. It seems to me that no matter how light your finger touches it, it will always detect as down. There seems to be no mechanism to prevent something like "spurious touch input". And, I tested it with some conductive objects (a key, the clip I used to short PWR pins on my desktop, a flash drive with metal USB port, a 3.5mm cable) and none of them seems to be able to trigger the touchpad. Maybe the touchpad doesn't work like a capacitive touchscreen? I've been using the quirk of 1:0:850 for a while and it seems just fine at the moment.

it seem that it using both capacitive(idk if I call it correct) and force-sensor based

you can test this on windows

0) enable hpatics feedback
1) any non-capacitive-object put on touchpad, just contact, no force put on it
2) you put the finger  on TP., just contact, no force put on it
3) you will notice that no click occured
4) release finger, and then do the 1) again but put force on it
5) do the same of 2)
6) you will notice that you dont have to put force, just contact, and it give click and haptic feedback

**Comment by NP-chaonay** [comment]
another test

1) do the above 0)
2) you just using that non-capacitive object to click >> dont work
3) do the above 2) and do this 2) >> it work

so this can conclude that it must have enough force and also capacitive in order to establish clicking

**Comment by m1nicrusher** [comment]
Hmmm, but even if I use a conductive metal and put force onto the touchpad it won't detect. The only non-finger object that works for me is a stick of sausage in the package. Regardless, I have been using 1:0:850 for a whole day and have seen absolutely no mis-touch or any unwanted behaviour. This is by far the closest experience as it is on Surface UEFI (I will try Windows someday)

**Comment by qzed** [comment]
@m1nicrusher can you check if everything behaves fine after suspend? We've seen some interaction between touchpad and touchscreen in the past.

**Comment by m1nicrusher** [comment]
> @m1nicrusher can you check if everything behaves fine after suspend? We've seen some interaction between touchpad and touchscreen in the past.

I've tried a few things:
- Suspend and wake up
- Suspend while touching the screen and wake up
- Suspend while using the pen and wake up
- Suspend while touching the touchpad and wake up

None of the above results in any malfunction. I remember on my Surface Pro 8 pen (or touchscreen?) used to stop working after suspend because it was in use while suspending, which will result in stuck status after wake up.

On SLS though, nothing goes wrong.

**Comment by m1nicrusher** [comment]
I do find on my SLS sometimes the keyboard stops working for a few seconds. Sometimes it will stuck on one key. For example when I'm typing, it will typeeeeeeeeeeeeeeeeeeeeeeeeeeee <- like that. Does any one have similar issue?

**Comment by m1nicrusher** [comment]
Update: I noticed an issue with my touchpad recently. Sometime it'll stop working for a second. Not sure what caused but I'm changing my quirk to 5:3:850 and I'll keep monitoring if this still happens.

**Comment by m1nicrusher** [comment]
I have updated the title of the issue since I don't see enough users using Ubuntu Unity and it doesn't use mainstream package to control touchpad as other distros.

**Comment by onkoe** [comment]
I don't think the keyboard and trackpad issues are related, so it may be wise to make a new issue for that instead.

That said, I also have the keyboard problem, but even after writing code for weeks on this thing, it hasn't hurt me once. At least it chooses good times to do it, like when updating packages 😊

**Comment by qzed** [comment]
Any chance there's any output in the dmesg log when that happens?

> At least it chooses good times to do it, like when updating packages

I guess that could mean that it's related to high cpu usage.

**Comment by m1nicrusher** [comment]
> I don't think the keyboard and trackpad issues are related, so it may be wise to make a new issue for that instead.

When the keyboard freezes on my SLS, the touchpad stops working; when the touchpad freezes, the keyboard won't work as well. If they are separate issues, at least here on my machine they happen at the same time (or at least one triggers another)

**Comment by m1nicrusher** [comment]
> > At least it chooses good times to do it, like when updating packages
> 
> I guess that could mean that it's related to high cpu usage.

The issue happens to me quite randomly. Sometimes just moving around the cursor but not really doing anything. Sometimes when scrolling down a webpage.

> Any chance there's any output in the dmesg log when that happens?

I've checked `dmesg` before and after the issue happens, nothing was added so I'm not sure if you can see any problems but here it is:
[dmesg.log](https://github.com/linux-surface/linux-surface/files/10473945/dmesg.log)

**Comment by m1nicrusher** [comment]
This has been getting very annoying. It literally happens almost every half an hour. I didn't notice this happening so frequent before the upgrading to the new* libinput package.

*: Some time before, you need to have the quirk for the touchpad to work. Then, at some point, touchpad works out of the box, you still need the quirk for some tweaks but at least it moves. I can't remember which version that was so I can't really do a downgrade to test...

**Comment by m1nicrusher** [comment]
The freezes are usually about 5+ seconds. So it's really kindda annoying to me, especially when it happens while I was typing. I'll have to spend extra time to delete the "kkkkkkkkkkkkkkkkkkkk" generated by the stucked key.

**Comment by m1nicrusher** [comment]
> One option is to enable debugging in the SAM driver, but that will cause a flood of messages... 

How to do that? It might be worth trying since `libinput measure` throws and exception when the freeze happens. I'm assuming the driver crashed somehow?

**Comment by qzed** [comment]
Okay, that is actually quite helpful.

> How to do that? It might be worth trying

Right. Because it freezes this long, we might actually have a good chance at figuring out a bit more about the issue. Can you run

```
echo "module surface_aggregator +p" | sudo tee /sys/kernel/debug/dynamic_debug/control
echo "module surface_hid +p" | sudo tee /sys/kernel/debug/dynamic_debug/control
```

Then monitor the dmesg log (best open a new terminal with `dmesg -w` running to the side). If you type something or use the touchpad, you should get a flood of messages with some containing `surface_serial_hub: ...` and others with actual data being transmitted.

Then try to reproduce the freeze. And once that happens, check if keyboard/touchpad still cause those messages to appear in the log.

This will give us an indication where the problem might be: I.e. if the EC isn't sending us events, or if the communication somehow fails, or if it's somewhere downstream of that.

> since `libinput measure` throws an exception when the freeze happens. I'm assuming the driver crashed somehow?

Oh, what kind of exception? I would assume that if it's some kind of comms failure or failure of the SAM driver internals, this should not happen.

Maybe you could also try running evtest on the keyboard or touchpad device (or both)? That could help figure out if there's something wrong with libinput.

**Comment by m1nicrusher** [comment]
> Then monitor the dmesg log (best open a new terminal with dmesg -w running to the side). 

I'm currently monitoring dmesg. It has a lot of information even when I'm not touching the laptop.

> Oh, what kind of exception? I would assume that if it's some kind of comms failure or failure of the SAM driver internals, this should not happen.

Those are the exceptions I posted before. Not sure if they are helpful.

```
Traceback (most recent call last):
  File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 432, in <module>
    main(sys.argv)
  File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 417, in main
    loop(device)
  File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 359, in loop
    for event in device. Events():
  File "/usr/lib/python3.11/site-packages/libevdev/device.py", line 541, in events
    raise EventsDroppedException()
libevdev.device.EventsDroppedException
```

**Comment by m1nicrusher** [comment]
@qzed OK, now it happened while I was monitoring dmesg. dmesg actually correctly responded my input during freeze. (e.g. size=27 while not touching anything; size=62 while touching the touchpad).

So I guess the problem is not the driver?

**Comment by qzed** [comment]
> So I guess the problem is not the driver?

At least not the receiver side... and it's also not a communication issue.

> Those are the exceptions I posted before. Not sure if they are helpful.
> 
> ```
> Traceback (most recent call last):
>   File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 432, in <module>
>     main(sys.argv)
>   File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 417, in main
>     loop(device)
>   File "/usr/libexec/libinput/libinput-measure-touchpad-pressure", line 359, in loop
>     for event in device. Events():
>   File "/usr/lib/python3.11/site-packages/libevdev/device.py", line 541, in events
>     raise EventsDroppedException()
> libevdev.device.EventsDroppedException
> ```

Ah, so it looks like the kernel sends a `EV_SYN` `SYN_DROPPED` event. As  far as I understand, this event is sent/inserted when the event queue gets full. Which should only happen if userspace opens a device and doesn't read from it quickly enough. I guess that means the SAM and surface_hid drivers should work as intended (since they handle lower-level stuff which eventually feeds into that queue), unless they're somehow spamming it.

Since you mentioned that there's a lot of activity even when you don't use the keyboard/touchpad, can you post an excerpt (a couple of seconds) from the log?

Can you run `evetest` on both the keyboard and touchpad device (make sure that events are emitted when you type / touch the touchpad) and see how they react when there's a freeze?

**Comment by m1nicrusher** [comment]
> Since you mentioned that there's a lot of activity even when you don't use the keyboard/touchpad, can you post an excerpt (a couple of seconds) from the log?

I will have to do it later tomorrow since the test requires Secure Boot to be turned off.

> Can you run evetest on both the keyboard and touchpad device (make sure that events are emitted when you type / touch the touchpad) and see how they react when there's a freeze?

There are 3 devices that look like the keyboard:
```
/dev/input/event16:	Microsoft Surface 045E:09AE
/dev/input/event17:	Microsoft Surface 045E:09AE Keyboard
/dev/input/event18:	Microsoft Surface 045E:09B0 Keyboard
```

I have chosen event17 since that is the first device that says "Keyboard". I noticed that when the keyboard/touchpad freezes, the test won't receive any input.

I'll keep doing some more tests later because it seems like when the tests are one, the freeze will be extremely difficult to trigger. While when I don't have the test on, sometimes it could trigger several times within a minute.

**Comment by qzed** [comment]
> I have chosen event17 since that is the first device that says "Keyboard". I noticed that when the keyboard/touchpad freezes, the test won't receive any input.

That would indicate it is somewhere in the kernel. You can verify that it's the correct device if it reacts to your input (when it's not frozen). There should be one for normal keyboard input and one for the touchpad input (and potentially more for function keys and other things).

**Comment by m1nicrusher** [comment]
> You can verify that it's the correct device if it reacts to your input (when it's not frozen)

Yes it does react. It poses a message when keydown, keyhold and keyup.

**Comment by m1nicrusher** [comment]
Also I have solved the Ubuntu Unity touchpad problem by forcing the system to use libinput (maybe?).

I followed steps on this link: https://askubuntu.com/questions/1156192/how-to-enable-tap-to-click-in-libinput-on-ubuntu-19-04-x11-unity

I created the following file at `/etc/X11/xorg.conf.d/30-touchpad.conf`:
```
Section "InputClass"
Identifier "touchpad"
Driver "libinput"
  MatchIsTouchpad "on"
  Option "Tapping" "on"
  Option "NaturalScrolling" "on"
  Option "ClickMethod" "clickfinger"
EndSection
```

Even the original post has a different issue on a different system, this magically works for me. By the way I haven't noticed any keyboard/touchpad freeze on Ubuntu Unity (kernel: 6.1.9-surface).

**Comment by m1nicrusher** [comment]
@qzed 

> Since you mentioned that there's a lot of activity even when you don't use the keyboard/touchpad, can you post an excerpt (a couple of seconds) from the log?

Here is a portion of the idle dmesg. It seems like we have len: 17, 15, 20 with no specific order.
[dmesg.log](https://github.com/linux-surface/linux-surface/files/10697706/dmesg.log)

**Comment by qzed** [comment]
Thanks, so it looks like some keyboard events (the ones with length 20 / instance ID 01; I assume you pressed a key during or slightly before the log?) but mostly it's sensor stuff (the ones with length 17 and 15 / Instance ID 06). So my guess is that this should be fine, but I'm still slightly worried because there are some keyboard events and all in all it seems a bit much communication.

**Comment by qzed** [comment]
Okay, so I can spot the following key events:
```
<clear> S S <clear>  G G <clear> <clear> Enter <clear> <clear>
```

It looks like there's about 100ms between repeated events. Also it kind of looks like it re-sends the key before sending the key-up/clear event. Or maybe also the clear event before the key-down event.

**Comment by m1nicrusher** [comment]
> I assume you pressed a key during or slightly before the log?

Well I definitely pressed Enter to initiate `dmesg -w > dmesg.log`, the S and G probably came from "dme**sg**"?

Nevertheless, the part I gave you is the very middle part of the test. I let it run for around 5 seconds and used Ctrl + C to stop it. During the test I didn't touch the keyboard/touchpad. What do you think?

**Comment by m1nicrusher** [comment]
I did the test again right after a fresh reboot. I do have IBus enabled and I noticed sometimes when it's switched to Chinese Intelligent Pinyin input method (but in English mode), keystrokes won't reflect on screen, and after I switched it back to English Australia (this probably disable IBus?) the previous keystrokes will then reflect on screen with a significant delay. I don't know if this have potentially caused anything but this time I did it without IBus enabled:

<details>
    <summary>dmesg again</summary>

```
[  320.582999] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  320.583003] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.583005] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.660670] surface_serial_hub serial0-0: rx: received data (size: 27)
[  320.660680] rx: 00000000: aa 55 00 11 00 e1 9c 1d 80 15 00 01 06 15 00 00
[  320.660684] rx: 00000010: a1 02 04 fc ff 4c fc b5 fe aa d4
[  320.660688] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  320.660692] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.660695] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.682129] surface_serial_hub serial0-0: rx: received data (size: 25)
[  320.682139] rx: 00000000: aa 55 00 0f 00 e2 9d 75 80 15 00 01 06 15 00 00
[  320.682143] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  320.682147] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  320.682151] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.682154] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.760407] surface_serial_hub serial0-0: rx: received data (size: 27)
[  320.760416] rx: 00000000: aa 55 00 11 00 e3 de 3d 80 15 00 01 06 15 00 00
[  320.760420] rx: 00000010: a1 02 04 fd ff 4c fc bb fe 05 b2
[  320.760423] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  320.760427] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.760429] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.781473] surface_serial_hub serial0-0: rx: received data (size: 25)
[  320.781483] rx: 00000000: aa 55 00 0f 00 e4 5b 15 80 15 00 01 06 15 00 00
[  320.781486] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  320.781489] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  320.781493] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.781496] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.881211] surface_serial_hub serial0-0: rx: received data (size: 27)
[  320.881221] rx: 00000000: aa 55 00 11 00 e5 18 5d 80 15 00 01 06 15 00 00
[  320.881224] rx: 00000010: a1 02 04 00 00 4b fc b3 fe d1 ef
[  320.881228] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  320.881232] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.881234] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.881390] surface_serial_hub serial0-0: rx: received data (size: 25)
[  320.881393] rx: 00000000: aa 55 00 0f 00 e6 19 35 80 15 00 01 06 15 00 00
[  320.881395] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  320.881397] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  320.881399] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.881401] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.958505] surface_serial_hub serial0-0: rx: received data (size: 27)
[  320.958516] rx: 00000000: aa 55 00 11 00 e7 5a 7d 80 15 00 01 06 15 00 00
[  320.958519] rx: 00000010: a1 02 04 fc ff 4c fc bd fe 03 5d
[  320.958523] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  320.958527] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.958530] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  320.979517] surface_serial_hub serial0-0: rx: received data (size: 25)
[  320.979526] rx: 00000000: aa 55 00 0f 00 e8 d7 d4 80 15 00 01 06 15 00 00
[  320.979529] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  320.979533] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  320.979537] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  320.979539] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.057801] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.057811] rx: 00000000: aa 55 00 11 00 e9 94 9c 80 15 00 01 06 15 00 00
[  321.057814] rx: 00000010: a1 02 04 00 00 4a fc b2 fe 54 aa
[  321.057818] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.057822] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.057825] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.078552] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.078562] rx: 00000000: aa 55 00 0f 00 ea 95 f4 80 15 00 01 06 15 00 00
[  321.078565] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.078569] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.078573] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.078575] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.156637] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.156647] rx: 00000000: aa 55 00 11 00 eb d6 bc 80 15 00 01 06 15 00 00
[  321.156651] rx: 00000010: a1 02 04 fd ff 4c fc b7 fe 68 f7
[  321.156654] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.156658] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.156661] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.177595] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.177606] rx: 00000000: aa 55 00 0f 00 ec 53 94 80 15 00 01 06 15 00 00
[  321.177609] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.177613] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.177618] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.177620] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.255993] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.256003] rx: 00000000: aa 55 00 11 00 ed 10 dc 80 15 00 01 06 15 00 00
[  321.256006] rx: 00000010: a1 02 04 fe ff 4c fc b5 fe ea 5f
[  321.256010] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.256013] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.256016] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.276709] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.276736] rx: 00000000: aa 55 00 0f 00 ee 11 b4 80 15 00 01 06 15 00 00
[  321.276740] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.276744] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.276748] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.276750] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.355171] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.355182] rx: 00000000: aa 55 00 11 00 ef 52 fc 80 15 00 01 06 15 00 00
[  321.355185] rx: 00000010: a1 02 04 00 00 4b fc b8 fe 2b 33
[  321.355189] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.355193] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.355195] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.376027] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.376037] rx: 00000000: aa 55 00 0f 00 f0 ee 47 80 15 00 01 06 15 00 00
[  321.376040] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.376044] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.376047] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.376050] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.454501] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.454511] rx: 00000000: aa 55 00 11 00 f1 ad 0f 80 15 00 01 06 15 00 00
[  321.454514] rx: 00000010: a1 02 04 fe ff 4c fc b7 fe 88 39
[  321.454518] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.454522] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.454525] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.475045] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.475056] rx: 00000000: aa 55 00 0f 00 f2 ac 67 80 15 00 01 06 15 00 00
[  321.475059] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.475063] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.475068] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.475071] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.553611] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.553621] rx: 00000000: aa 55 00 11 00 f3 ef 2f 80 15 00 01 06 15 00 00
[  321.553624] rx: 00000010: a1 02 04 00 00 4c fc b4 fe 6b 27
[  321.553628] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.553632] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.553634] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.574189] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.574198] rx: 00000000: aa 55 00 0f 00 f4 6a 07 80 15 00 01 06 15 00 00
[  321.574202] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.574206] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.574210] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.574212] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.652144] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.652153] rx: 00000000: aa 55 00 11 00 f5 29 4f 80 15 00 01 06 15 00 00
[  321.652156] rx: 00000010: a1 02 04 ff ff 4b fc ba fe 59 5b
[  321.652160] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.652164] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.652166] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.673268] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.673277] rx: 00000000: aa 55 00 0f 00 f6 28 27 80 15 00 01 06 15 00 00
[  321.673280] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.673284] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.673287] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.673290] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.751488] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.751499] rx: 00000000: aa 55 00 11 00 f7 6b 6f 80 15 00 01 06 15 00 00
[  321.751502] rx: 00000010: a1 02 04 ff ff 4b fc b3 fe c1 e1
[  321.751506] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.751510] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.751513] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.772657] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.772667] rx: 00000000: aa 55 00 0f 00 f8 e6 c6 80 15 00 01 06 15 00 00
[  321.772670] rx: 00000010: 89 02 04 04 00 00 00 07 b6
[  321.772674] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 15)
[  321.772678] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.772681] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.850818] surface_serial_hub serial0-0: rx: received data (size: 27)
[  321.850827] rx: 00000000: aa 55 00 11 00 f9 a5 8e 80 15 00 01 06 15 00 00
[  321.850830] rx: 00000010: a1 02 04 fd ff 4a fc ba fe ad a6
[  321.850833] surface_serial_hub serial0-0: rx: parser: valid frame found (type: 0x00, len: 17)
[  321.850837] surface_serial_hub serial0-0: rx: parser: valid command found (tc: 0x15, cid: 0x00)
[  321.850840] surface_serial_hub serial0-0: rtl: handling event (rqid: 0x0015)
[  321.871795] surface_serial_hub serial0-0: rx: received data (size: 25)
[  321.871806] rx: 00000000: aa 55 00 0f 00 fa a4 e6 80 15 00 01 06 15 00 00
[  321.871809] rx: 00000010: 89 02 04 04 00 00 00 07 b6
```

<details>

**Comment by m1nicrusher** [comment]
Basically I have two input methods installed:
![image](https://user-images.githubusercontent.com/19246983/217980174-980b8c93-d741-4bee-bb34-00c4d8de79c5.png)
I can either switch to English Australia so that I can type purely English, or I can switch to Chinese and use IBus to type in Chinese but it also provides an English mode so that you don't have to switch out from Chinese input method to type English. So you simply press <kbd>Shift</kbd> once, and it will switch to English mode (note you're still in Chinese input method but you can now type English). It normally works the same way as English Australia, except for the bug that I mentioned above. I'm assuming this is a bug of IBus and it might not be related to the keyboard issue.

**Comment by qzed** [comment]
> Well I definitely pressed Enter to initiate `dmesg -w > dmesg.log`, the S and G probably came from "dme**sg**"?

Yeah, I assumed you typed `dmesg <enter>` at some point.

> It's still 15 and 17 so I assume it's still S and G (but without Enter)?

As I've mentioned before, the length 15 and 17 events seem to be sensor events. Length 20 are key input events.

Actually, this time I can't spot any key events. You can check this by looking at the `80 15 00 01 06 15 00 00` part (usually the last part in the first `rx` line). More or less the only thing that's flexible for input events on the SLS is the `06`. This is the instance ID, which indicates from which subdevice the event originates. `01` means keyboard, `02` the pen holder, `03` the touchpad, `06` the sensors, and `07` UCM UCSI. There's a bit more on the protocol layout here: https://www.kernel.org/doc/html/latest/driver-api/surface_aggregator/ssh.html (the `80` is essentially the start of the SSH command structure).

Also: In case the first line lines up nicely like this, the second `rx` line starts with the payload (and ends with the two CRC bytes). So in your earlier log, you had a payload of `01 00 16 00 00 00 00 00 00 00 00 00` for the keyboard input command. and the hex `16` is the key code for [`S`](https://gist.github.com/MightyPork/6da26e382a7ad91b5496ee55fdc73db2#file-usb_hid_keys-h-L58).

> I do have IBus enabled and I noticed sometimes when it's switched to Chinese Intelligent Pinyin input method (but in English mode), keystrokes won't reflect on screen, and after I switched it back to English Australian (this probably disable IBus?) the previous keystrokes will then reflect on screen with a significant delay.

Unfortunately, I'm not really familiar with this. But from what I gather, it runs entirely in user-space, so it shouldn't be able to effect the kernel in any way. My guess is that any delay you experience when switching is probably related to IBus itself.

**Comment by qzed** [comment]
Regarding things to check and try out: When the keyboard freezes and you continue to type, do the messages with `len: 20` still show up in normal frequency? Or better yet, even thought it might be a bit of data, can you try to collect a log with this debug info from the freeze and annotate it in which second (roughly) the freeze started and when it stopped. And ideally, could you try and check that without using the touchpad so it's contained to the keyboard only for now?

**Comment by m1nicrusher** [comment]
> Regarding things to check and try out: When the keyboard freezes and you continue to type, do the messages with `len: 20` still show up in normal frequency? 

I assume yes. The last time I did the experiment was days ago and I find it difficult to trigger the freeze while it's showing the log.

> Or better yet, even thought it might be a bit of data, can you try to collect a log with this debug info from the freeze and annotate it in which second (roughly) the freeze started and when it stopped. 

I can definitely try this, but as I said, it somehow is hard to trigger the freeze when the log is running...

> And ideally, could you try and check that without using the touchpad so it's contained to the keyboard only for now?

Yes absolutely.

**Comment by qzed** [comment]
> I can definitely try this, but as I said, it somehow is hard to trigger the freeze when the log is running...

No rush, if it makes it easier, you can also try to use the touchpad. It just creates more events for us to analyze. I guess I should write a script for that anyways...

**Comment by m1nicrusher** [comment]
> Regarding things to check and try out: When the keyboard freezes and you continue to type, do the messages with `len: 20` still show up in normal frequency?

I think in this case yes. I've attached a huge log here.

Before the timestamp `385317.00`, it is 100% working. After at around `385317.60`, it is 100% frozen. 

In between, I'm not sure cuz it happened too quick.

At roughly `385320.06`, it recovered.

[dmesg.log](https://github.com/linux-surface/linux-surface/files/10732729/dmesg.log)

**Comment by m1nicrusher** [comment]
I do notice something that is after it freezes, everything you typed within the frozen period will be reflected after it is recovered.

For example in my cases above while I was navigating I noticed my touchpad doesn't work anymore, so I typed "aaaaaaaaa" (<kbd>A</kbd> key 9 times) and after it recovers control, the 9 letter "a" appeared in my text box immediately.

**Comment by qzed** [comment]
I've pushed a script for converting the dmesg log to json to https://github.com/linux-surface/surface-aggregator-module/blob/master/scripts/dmesg/dmesg_parse.py. Currently looking into the keyboard events.

**Comment by qzed** [comment]
Okay, here's a list of keyboard events from the log, with timestamps:

```json
{"key": "a",       "time":385317.900401}
{"key": "<clear>", "time":385317.985508}
{"key": "<clear>", "time":385318.085402}
{"key": "a",       "time":385318.095134}
{"key": "<clear>", "time":385318.170918}
{"key": "<clear>", "time":385318.270480}
{"key": "a",       "time":385318.270485}
{"key": "<clear>", "time":385318.341195}
{"key": "<clear>", "time":385318.441310}
{"key": "a",       "time":385318.444665}
{"key": "<clear>", "time":385318.526140}
{"key": "<clear>", "time":385318.624963}
{"key": "a",       "time":385318.651312}
{"key": "<clear>", "time":385318.723401}
{"key": "<clear>", "time":385318.823142}
{"key": "a",       "time":385318.826907}
{"key": "<clear>", "time":385318.906745}
{"key": "<clear>", "time":385319.006185}
{"key": "a",       "time":385319.054814}
{"key": "<clear>", "time":385319.122179}
{"key": "<clear>", "time":385319.221700}
{"key": "<clear>", "time":385319.321060}
{"key": "a",       "time":385319.395634}
{"key": "<clear>", "time":385319.458480}
{"key": "<clear>", "time":385319.557632}
{"key": "<clear>", "time":385319.656849}
{"key": "<clear>", "time":385319.756129}
{"key": "<clear>", "time":385319.855440}
{"key": "<clear>", "time":385319.955008}
{"key": "a",       "time":385320.040022}
{"key": "<clear>", "time":385320.114361}
{"key": "a",       "time":385320.207777}
{"key": "<clear>", "time":385320.288008}
{"key": "<clear>", "time":385320.386995}
{"key": "<clear>", "time":385320.486496}
{"key": "<clear>", "time":385320.586096}
{"key": "<clear>", "time":385320.686199}
{"key": "<clear>", "time":385320.785766}
{"key": "<clear>", "time":385320.885168}
{"key": "<clear>", "time":385320.984465}
{"key": "<clear>", "time":385321.084418}
{"key": "<clear>", "time":385321.185104}
{"key": "ctrl",    "time":385323.670273}
{"key": "ctrl+c",  "time":385323.728015}
```

So there are definitely key events in the time period you describe. Meaning we can rule out any issues on the SAM/EC side and the issue has to be somewhere in the kernel driver stack. Unfortunately, that doesn't really give us an indication where exactly the problem in the kernel is since the logs come essentially directly from the receiving part. So it could be in the SAM driver event handling, in the Surface HID driver, or in the HID and input stack.

So I guess the next step would be adding a debug print before https://github.com/linux-surface/surface-aggregator-module/blob/master/module/src/clients/surface_hid.c#L164 (i.e. add that to the module, build and load it). That's the hand-off point of the SAM / Surface HID drivers to the HID / input subsystem. Let me know if you need assistance/details for that.

**Comment by m1nicrusher** [comment]
> So I guess the next step would be adding a debug print before

Oh, that would be something I've never done before...

I do remember back in the days when I first bought this laptop it didn't have this issue at all, other than sometimes touchpad is weird (which is solved by using better quirk arguments).

**Comment by qzed** [comment]
Okay, so in short what you want to do is:
1. clone the module and cd into it
```sh
git clone https://github.com/linux-surface/surface-aggregator-module`
cd surface-aggregator-module/module
```
2. Add the debug print, e.g.
```diff
diff --git a/module/src/clients/surface_hid.c b/module/src/clients/surface_hid.c
index e8f74a0..4ac25cd 100644
--- a/module/src/clients/surface_hid.c
+++ b/module/src/clients/surface_hid.c
@@ -161,6 +161,9 @@ static u32 ssam_hid_event_fn(struct ssam_event_notifier *nf, const struct ssam_e
 	if (event->command_id != 0x00)
 		return 0;
 
+	dev_info(shid->dev, "%s: iid=%u\n", __func__, event->instance_id);
+	print_hex_dump(KERN_INFO, "hid rx: ", DUMP_PREFIX_OFFSET, 16, 1, &event->data[0], event->length, false);
+
 	hid_input_report(shid->hid, HID_INPUT_REPORT, (u8 *)&event->data[0], event->length, 0);
 	return SSAM_NOTIF_HANDLED;
 }
```

3. build the module
```sh
make clean all -j$(nproc)
```

4. load the module
```sh
make modprobe-unload   # unload the built-in modules
make insmod            # load new modules
```

For the last step, you will need an external keyboard.

After that, test with the modified modules loaded (no need to enable debug logging) and try to reproduce the issue. On each input event, you should see a message printed to the dmesg log.

**Comment by qzed** [comment]
@tyush It looks like `platform_profile` has recently been moved to a module. I've updated the script to load that as well, so can you update (`git pull`) the module? Alternatively, you can run `modprobe platform_profile` manually before running `make insmod`.

**Comment by tyush** [comment]
Yep, that seems to work!

Will report back when I experience a freeze.

**Comment by tyush** [comment]
@qzed Got a [log](https://gist.github.com/tyush/2b38323d5784750300039ce9bbe9d823) from a freeze. Freeze occurred around timestamp 4572.

**Comment by qzed** [comment]
How long did it freeze for, approximately?

**Comment by tyush** [comment]
I'd say about 3-4 seconds.

**Comment by qzed** [comment]
Alright, so that checks out. There's a gap in the log from 4572.370264 to 4575.513245 (~3s).
```
[ 4572.370264] hid rx: 00000020: 00 00 00 00 31 0c 01 00 7b 01 91 1e
[ 4575.513245] surface_hid 01:15:01:03:00: ssam_hid_event_fn: iid=3
```
So it looks like it's internal to the SAM modules. Somewhere between receiving the event from the EC and the HID driver receiving the data via the workqueue.

I'll have to check what that could be... unfortunately that might not be that easy to figure out, especially if it's an issue with the workqueue itself.

**Comment by qzed** [comment]
Sorry for the long silence. Can you change and try the following?
```diff
diff --git a/module/src/controller.c b/module/src/controller.c
index bb11ef5..758872d 100644
--- a/module/src/controller.c
+++ b/module/src/controller.c
@@ -825,7 +825,7 @@ static int ssam_cplt_init(struct ssam_cplt *cplt, struct device *dev)
 
        cplt->dev = dev;
 
-       cplt->wq = create_workqueue(SSAM_CPLT_WQ_NAME);
+       cplt->wq = alloc_workqueue(SSAM_CPLT_WQ_NAME, WQ_UNBOUND | WQ_MEM_RECLAIM | WQ_HIGHPRI, 0);
        if (!cplt->wq)
                return -ENOMEM;
 
```

This changes some parameters of the work-queue handling EC events, including input events. In particular, it
- Bumps up the priority of tasks executed on the queue.
- Allows processing of multiple "work structs" in parallel. This means each event category should now be processed in parallel, which in turn means that input events can no longer be blocked by e.g. battery events. So if there's some event handling function in some other category misbehaving and blocking, it should now no longer affect input events.

Depending on the root cause of the problem, either one by itself may be enough.

See https://github.com/linux-surface/linux-surface/issues/1026#issuecomment-1440450678 on how to recompile and load the module with those changes.

**Comment by tyush** [comment]
Don't worry about it; I understand.

I've loaded the module with those changes. I'll report back when I experience an input hang, or if at all.

**Comment by tyush** [comment]
4200 seconds of usage in, and usually I would've experienced 4-5 freezes in that time. I'd say it worked! Trying without WQ_HIGHPRI right now.

**Comment by tyush** [comment]
Few workdays running without WQ_HIGHPRI. No freezes experienced in that time. Trying with only WQ_HIGHPRI and WQ_UNBOUND now.

**Comment by qzed** [comment]
The next releases (v6.2.14 / v6.2.13-2) contain the change above, so hopefully that will fix the issue.

**Comment by qzed** [comment]
Any issues with this on the latest kernels or can I send this upstream and mark it as fixed?

**Comment by tyush** [comment]
I haven't encountered the issue in about 2 weeks since updating.

**Comment by qzed** [comment]
Thanks!

**Comment by qzed** [comment]
I've sent the patch upstream.
